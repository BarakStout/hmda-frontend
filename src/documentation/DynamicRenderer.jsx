import React, { useState, useEffect } from 'react'
import Markdown from 'markdown-to-jsx'
import { Link } from 'react-router-dom'
import LoadingIcon from '../common/LoadingIcon.jsx'
import NotFound from '../common/NotFound.jsx'
import { getMarkdownUrl, isBadSlug } from './markdownUtils'

const DynamicRenderer = props => {
  const [data, setData] = useState(null)
  const { year, slug } = props
  const badSlug = isBadSlug(year, slug)


  useEffect(function (){
    if(badSlug) return
    fetch(getMarkdownUrl(year, slug)).then(res => {
      res.text().then(setData)
    })
  }, [year, slug, badSlug])

  useEffect(function (){
    if(!data || badSlug) return
    const { hash } = window.location
    if(hash) {
      setTimeout(() => {
        const stripped = hash.replace(/[#_]/g, '')
        const id = stripped + stripped
        const element = document.getElementById(id)
        if (element) setTimeout(()=>element.scrollIntoView(), 0)
      }, 0)
    }
  })

  if(badSlug) return <NotFound/>

  return (
    <div className="Markdown-Wrapper">
    <Link className="BackLink" to={`/documentation/${year}`}>{'\u2b05'} {year} DOCUMENTATION</Link>
    {data
      ? <Markdown>{data}</Markdown>
      : <LoadingIcon/>
    }
    </div>
  )
}

export default DynamicRenderer
