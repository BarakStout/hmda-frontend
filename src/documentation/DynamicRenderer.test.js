import React from 'react'
import DynamicRenderer from './DynamicRenderer'
import renderer from 'react-test-renderer'
import { BrowserRouter } from 'react-router-dom'

describe('DynamicRenderer', () => {
  it('renders plain text consistently', () => {
    const tree = renderer
      .create(
        <BrowserRouter>
          <DynamicRenderer year='2018' slug='identifiers-faq' />
        </BrowserRouter>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders tables consistently', () => {
    const tree = renderer
      .create(
        <BrowserRouter>
          <DynamicRenderer year='2018' slug='modified-lar-schema' />
        </BrowserRouter>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders lists consistently', () => {
    const tree = renderer
      .create(
        <BrowserRouter>
          <DynamicRenderer year='2018' slug='derived-data-fields' />
        </BrowserRouter>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders images consistently', () => {
    const tree = renderer
      .create(
        <BrowserRouter>
          <DynamicRenderer year='2019' slug='annual-filing-dates' />
        </BrowserRouter>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
