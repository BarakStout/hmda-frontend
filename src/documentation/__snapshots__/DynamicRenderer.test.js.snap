// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DynamicRenderer renders images consistently 1`] = `
<div
  className="Markdown-Wrapper"
>
  <a
    className="BackLink"
    href="/documentation/2019"
    onClick={[Function]}
  >
    ⬅
     
    2019
     DOCUMENTATION
  </a>
  <div
    className="LoadingIconWrapper"
  >
    <div
      className="LoadingIcon"
    />
  </div>
</div>
`;

exports[`DynamicRenderer renders lists consistently 1`] = `
<div
  className="Markdown-Wrapper"
>
  <a
    className="BackLink"
    href="/documentation/2018"
    onClick={[Function]}
  >
    ⬅
     
    2018
     DOCUMENTATION
  </a>
  <div
    className="LoadingIconWrapper"
  >
    <div
      className="LoadingIcon"
    />
  </div>
</div>
`;

exports[`DynamicRenderer renders plain text consistently 1`] = `
<div
  className="Markdown-Wrapper"
>
  <a
    className="BackLink"
    href="/documentation/2018"
    onClick={[Function]}
  >
    ⬅
     
    2018
     DOCUMENTATION
  </a>
  <div
    className="LoadingIconWrapper"
  >
    <div
      className="LoadingIcon"
    />
  </div>
</div>
`;

exports[`DynamicRenderer renders tables consistently 1`] = `
<div
  className="Markdown-Wrapper"
>
  <a
    className="BackLink"
    href="/documentation/2018"
    onClick={[Function]}
  >
    ⬅
     
    2018
     DOCUMENTATION
  </a>
  <div
    className="LoadingIconWrapper"
  >
    <div
      className="LoadingIcon"
    />
  </div>
</div>
`;
